nav2_tasks/include/nav2_tasks/goal_reached_condition.hpp

```diff
@@ -0,0 +1,96 @@
+// Copyright (c) 2019 Intel Corporation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef NAV2_TASKS__GOAL_REACHED_CONDITION_HPP_
+#define NAV2_TASKS__GOAL_REACHED_CONDITION_HPP_
+
+#include <string>
+#include <memory>
+#include "rclcpp/rclcpp.hpp"
+#include "behaviortree_cpp/condition_node.h"
+#include "nav2_robot/robot.hpp"
+#include "geometry_msgs/msg/pose_with_covariance_stamped.hpp"
+#include "nav2_tasks/compute_path_to_pose_task.hpp"
+
+namespace nav2_tasks
+{
+
+class GoalReachedCondition : public BT::ConditionNode
+{
+public:
+  explicit GoalReachedCondition(const std::string & condition_name)
+  : BT::ConditionNode(condition_name), initialized_(false)
+  {
+  }
+
+  GoalReachedCondition() = delete;
+
+  ~GoalReachedCondition()
+  {
+  }
+
+  BT::NodeStatus tick() override
+  {
+    if (!initialized_) {
+      node_ = blackboard()->template get<rclcpp::Node::SharedPtr>("node");
+
+      node_->get_parameter_or<double>("goal_reached_tol", goal_reached_tol_, 0.25);
+
+      robot_ = std::make_unique<nav2_robot::Robot>(node_);
+
+      initialized_ = true;
+    }
+
+    if (goalReached()) {
+      return BT::NodeStatus::SUCCESS;
+    }
+    return BT::NodeStatus::FAILURE;
+  }
+
+  bool
+  goalReached()
+  {
+    auto current_pose = std::make_shared<geometry_msgs::msg::PoseWithCovarianceStamped>();
+
+    if (!robot_->getCurrentPose(current_pose)) {
+      RCLCPP_DEBUG(node_->get_logger(), "Current robot pose is not available.");
+      return false;
+    }
+
+    blackboard()->get<nav2_tasks::ComputePathToPoseCommand::SharedPtr>("goal", command_); 
+    double dx = command_->pose.position.x - current_pose->pose.pose.position.x;
```

Comment:
Can you replace this with a function, or add a TODO to replace it?

There is a function `dwb_core.cpp` that seems a candidate `getSquareDistance`. However, this works with `Pose2D`s instead of `Pose`. I'm inclined to try to make it a templated function so it works with either type.

Reply:
Added TODO.